slasher/rpc/detect_update_min_max_span.go

```diff
@@ -0,0 +1,135 @@
+package rpc
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/pkg/errors"
+	ethpb "github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+// Detector is an interface used to implement the slashing surround an surrounded detection
+// methods.
+type Detector interface {
+	Detect(attestationEpochSpan uint64, recorderEpochSpans *ethpb.MinMaxEpochSpan, attestationSourceEpoch uint64) uint64
+}
+
+// maxDetector is a detector used to detect surround attestations.
+type maxDetector struct{}
+
+// Detect max detector is a detector used to detect surround attestations.
+func (d maxDetector) Detect(attestationEpochSpan uint64, recorderEpochSpans *ethpb.MinMaxEpochSpan, attestationSourceEpoch uint64) uint64 {
+	maxSpan := uint64(recorderEpochSpans.MaxEpochSpan)
+	if maxSpan > attestationEpochSpan {
+		return maxSpan + attestationSourceEpoch
+	}
+	return 0
+}
+
+// minDetector is a detector used to detect surrounded attestations.
+type minDetector struct{}
+
+// Detect min detector is a detector used to detect surrounded attestations.
+func (d minDetector) Detect(attestationEpochSpan uint64, recorderEpochSpans *ethpb.MinMaxEpochSpan, attestationSourceEpoch uint64) uint64 {
+	minSpan := uint64(recorderEpochSpans.MinEpochSpan)
+	if minSpan < attestationEpochSpan {
+		return minSpan + attestationSourceEpoch
+	}
+	return 0
+}
+
+// DetectAndUpdateMaxEpochSpan is used to detect and update the max span of an incoming attestation.
+// max span is the span between the current attestation source epoch and the furthest attestation
+// target that its source epoch is lower then it.
+//
+// logic is following the detection method designed by https://github.com/protolambda
+// from here: https://github.com/protolambda/eth2-surround/blob/master/README.md#min-max-surround
+func (ss *Server) DetectAndUpdateMaxEpochSpan(ctx context.Context, source uint64, target uint64, validatorIdx uint64) (uint64, error) {
+	targetEpoch, span, spanMap, err := ss.detectSlashingByEpochSpan(source, target, validatorIdx, maxDetector{})
+	if err != nil {
+		return 0, err
+	}
+	if targetEpoch > 0 {
+		return targetEpoch, nil
+	}
+	for i := uint64(1); i < target-source; i++ {
+		val := uint32(span - i - 1)
+		if _, ok := spanMap.EpochSpanMap[source+i]; !ok {
+			spanMap.EpochSpanMap[source+i] = &ethpb.MinMaxEpochSpan{}
+		}
+		if spanMap.EpochSpanMap[source+i].MaxEpochSpan < val {
+			spanMap.EpochSpanMap[source+i].MaxEpochSpan = val
+		} else {
```

Comment:
I dont think you need this else

Reply:
i want to stop updating values if i find equal or higher values

Reply:
I see it now!